sklearn/metrics/_ranking.py

```diff
@@ -1487,20 +1495,27 @@ def _dcg_sample_scores(y_true, y_score, k=None, log_base=2, ignore_ties=False):
         Cumulative Gain (the DCG obtained for a perfect ranking), in order to
         have a score between 0 and 1.
     """
-    discount = 1 / (np.log(np.arange(y_true.shape[1]) + 2) / np.log(log_base))
+    xp, _, device = get_namespace_and_device(y_true, y_score)
+    max_float_dtype = _max_precision_float_dtype(xp, device)
```

Comment:
Similar question as in #30878, if we should be using max precision for metrics, lets make this consistent for all metrics and move away from `_find_matching_floating_dtype`.